lib/RedisConnection.js

```diff
@@ -47,42 +47,52 @@ RedisConnection.prototype.isAvailable = function () {
 
 /** @override */
 RedisConnection.prototype.set = function (key, val, maxAgeMs, setWhenNotExist) {
-  var deferred = Q.defer()
-  var params = [key, this._compress(val), 'PX', maxAgeMs]
-  if (setWhenNotExist) params.push('NX')
-  this._client.set(params, this._makeNodeResolverWithTimeout(deferred, 'set', 'Redis [set] key: ' + key))
-  return deferred.promise
+  return this._compress(val)
+  .thenBound(function (compressedVal) {
+    var params = [key, compressedVal, 'PX', maxAgeMs]
+    if (setWhenNotExist) params.push('NX')
+
+    var deferred = Q.defer()
+    this._client.set(params, this._makeNodeResolverWithTimeout(deferred, 'set', 'Redis [set] key: ' + key))
+    return deferred.promise
+  }, this)
 }
 
 /** @override */
 RedisConnection.prototype.mset = function (items, maxAgeMs, setWhenNotExist) {
   if (!items || !items.length) return Q.resolve(undefined)
 
-  var deferred = Q.defer()
-  var commands = []
-  var i, l
-  if (setWhenNotExist) {
-    // Use "SET" to set each key with a "NX" flag.
-    for (i = 0, l = items.length; i < l; i++) {
-      commands.push(['set', items[i].key, this._compress(items[i].value), 'PX', maxAgeMs, 'NX'])
-    }
-  } else {
-    // Use "MSET" to set all the keys and "EXPIRE" to set TTL for each key
-    var msetCommand = ['MSET']
-    commands.push(msetCommand)
-    for (i = 0, l = items.length; i < l; i++) {
-      var key = items[i].key
-      // Append key value arguments to the set command.
-      msetCommand.push(key, this._compress(items[i].value))
-      // Append an expire command.
-      commands.push(['EXPIRE', key, Math.floor(maxAgeMs / 1000)])
+  var compressedPromises = items.map(function (item) {
```

Comment:
It doesn't change what this function does, but it changes the order.

In the old code, we compress everything up-front, but they are all done in one single computation unit.

In the new code, the computation is split into smaller pieces (assuming snappy is smart enough to do that), so we can interleave the call to Redis in the gaps for the items that finish compressing earlier than others. The total amount of time to cache a set of keys might be shorter.

Admittedly, multiple Redis calls will introduce overhead too, so I'm not sure if it's worth it.